arm7/lib/src/SND_seq.c

```diff
@@ -0,0 +1,1323 @@
+            case SND_INST_PSG:
+                // only channels 8, 9, 10, 11, 12, 13 support PSG
+                allowedChannels = 0x3F00;
+                break;
+            case SND_INST_NOISE:
+                // only channels 14 and 15 support noise
+                allowedChannels = 0xC000;
+                break;
+            default:
+                return;
+        }
+
+        allowedChannels &= track->channelMask;
+
+        chn = SND_AllocExChannel(allowedChannels,
+            player->prio + track->priority,
+            track->flags.channelMask,
+            ChannelCallback,
+            track);
+        if (chn == NULL)
+            return;
+
+        if (!SND_NoteOn(
+                chn, midiKey, velocity, track->flags.tie ? -1 : length, player->bank, &inst))
+        {
+            chn->priority = 0;
+            SND_FreeExChannel(chn);
+            return;
+        }
+
+        chn->channelLLNext = track->channelLLHead;
+        track->channelLLHead = chn;
+    }
+
+    if (track->envAttack != 0xFF)
+        SND_SetExChannelAttack(chn, track->envAttack);
+
+    if (track->envDecay != 0xFF)
+        SND_SetExChannelDecay(chn, track->envDecay);
+
+    if (track->envSustain != 0xFF)
+        SND_SetExChannelSustain(chn, track->envSustain);
+
+    if (track->envRelease != 0xFF)
+        SND_SetExChannelRelease(chn, track->envRelease);
+
+    chn->sweepPitch = track->sweepPitch;
+    if (track->flags.portamento)
+        chn->sweepPitch += (short)((track->portamentoKey - midiKey) << 6);
+
+    if (track->portamentoTime == 0)
+    {
+        chn->sweepLength = length;
+        chn->flags.autoSweep = FALSE;
+    }
+    else
+    {
+        int swp = track->portamentoTime * track->portamentoTime;
+        swp *= chn->sweepPitch < 0 ? -chn->sweepPitch : chn->sweepPitch;
+        swp >>= 11;
+        chn->sweepLength = swp;
+    }
+
+    chn->sweepCounter = 0;
+}
+
+static int TrackStepTicks(
+    struct SNDTrack *track, struct SNDPlayer *player, int trackIdx, u32 playNotes)
+{
+    (void)trackIdx;
+
+    struct SNDExChannel *chn;
+    u8 cmd;
+    enum SNDSeqVal valueType;
+    BOOL specialValueType;
+    BOOL runCmd;
+    s32 length;
+    int midiKey;
+
+    s32 par;
+
+    for (chn = track->channelLLHead; chn; chn = chn->channelLLNext)
+    {
+        if (chn->length > 0)
+            chn->length--;
+
+        if (!chn->flags.autoSweep && chn->sweepCounter < chn->sweepLength)
+            chn->sweepCounter++;
+    }
+
+    if (track->flags.noteFinishWait)
+    {
+        if (track->channelLLHead)
+            return 0;
+        track->flags.noteFinishWait = FALSE;
+    }
+
+    if (track->wait > 0)
+    {
+        track->wait--;
+        if (track->wait > 0)
+            return 0;
+    }
+
+    SeqCacheFetch(track->cur);
+
+    while (track->wait == 0 && !track->flags.noteFinishWait)
+    {
+        specialValueType = FALSE;
+        runCmd = TRUE;
+
+        cmd = TrackReadU8(track);
+
+        if (cmd == 0xA2)
```

Comment:
should cmds have constants? I imagine the list may get kinda long if they do

Reply:
I've thought about it and then I kinda realized those constants would be used in that single case only. Maybe still a good idea, but I for me found it easier to actually see which byte represents which functionality (though adding a comment to summarize functionality might be a good idea).